File: AlgorithmsPackPythonSolutions/47. Permutations II.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        result = []
        filter_set = set()
        self.backtrack(nums, 0, filter_set)
        for permutation in filter_set:
            result.append(list(permutation))
        return result

    def backtrack(self, nums, index, filter_set):
        if index == len(nums):
            filter_set.add(tuple(nums))
            return

        for i in range(index, len(nums)):
            nums[index], nums[i] = nums[i], nums[index]
            self.backtrack(nums, index + 1, filter_set)
            nums[index], nums[i] = nums[i], nums[index]

File: AlgorithmsPackJavaSolutions/47. Permutations II.java
class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        Set<List<Integer>> filterSet = new HashSet<>();
        backtrack(nums, 0, filterSet);
        for (List<Integer> permutation : filterSet) {
            result.add(permutation);
        }
        return result;
    }

    private void backtrack(int[] nums, int index, Set<List<Integer>> filterSet) {
        if (index == nums.length) {
            List<Integer> permutation = new ArrayList<>();
            for (int num : nums) {
                permutation.add(num);
            }
            filterSet.add(permutation);
            return;
        }

        for (int i = index; i < nums.length; i++) {
            swap(nums, i, index);
            backtrack(nums, index + 1, filterSet);
            swap(nums, i, index);
        }
    }

    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}
